x-pack/plugins/lens/public/app_plugin/share_action.ts

```diff
@@ -45,8 +45,7 @@ function getShareURLForSavedObject(
   );
 }
 
-function getShortShareableURL(
-  shortUrlService: (params: LensAppLocatorParams) => Promise<string>,
+export function getLocatorParams(
   data: LensAppServices['data'],
```

Comment:
Because of the way URLs are generate for saved visualizations and unsaved, the URL format are pretty different (explicit vs short URLs).
Due to this difference the `shareableUrlForSavedObject` property has been added and used also for the reporting when avaialble: https://github.com/elastic/kibana/pull/153429/files#diff-be49f342798d67aec36153234bac3806d1b22566c1cd71a84b95be7c77157cc4R121

Reply:
I'm sorry, I meant for my comment to be specific only to the `locatorParams`, and not the shareableUrl. The locator params are what actually gets used as the app state for the Reporting API. 

To clarify: the request used to generate a PNG report from Lens, and the URL seen when you click "Copy POST URL", looks like:
`http://localhost:5601/api/reporting/generate/pngV2?jobParams=%28browserTimezone%3AAmerica%2FPhoenix%2Clayout%3A%28dimensions%3A%28height%3A593%2Cwidth%3A1171.7421875%29%2Cid%3Apreserve_layout%29%2ClocatorParams%3A%28id%3ALENS_APP_LOCATOR%2Cparams%3A%28activeDatasourceId%3AformBased%2CdataViewSpecs%3A%21%28%29%2CdatasourceStates%3A%28formBased%3A%28isLoading%3A%21f%2Cstate%3A%28currentIndexPatternId%3Aff959d40-b880-11e8-a6d9-e546fe2bba5f%2Clayers%3A%28%2797c63ea6-9305-4755-97d1-0f26817c6f9a%27%3A%28columnOrder%3A%21%28%279f61a7df-198e-4754-b34c-81ed544136ba%27%2Cebcb19af-0900-4439-949f-d8cd9bccde19%2C%275575214b-7f21-4b6c-8bc1-34433c6a0c58%27%29%2Ccolumns%3A%28%275575214b-7f21-4b6c-8bc1-34433c6a0c58%27%3A%28dataType%3Anumber%2CisBucketed%3A%21f%2Clabel%3A%27Count%20of%20records%27%2CoperationType%3Acount%2Cscale%3Aratio%2CsourceField%3A___records___%29%2C%279f61a7df-198e-4754-b34c-81ed544136ba%27%3A%28dataType%3Astring%2CisBucketed%3A%21t%2Clabel%3A%27Top%20values%20of%20category.keyword%27%2CoperationType%3Aterms%2Cparams%3A%28missingBucket%3A%21f%2CorderBy%3A%28columnId%3A%275575214b-7f21-4b6c-8bc1-34433c6a0c58%27%2Ctype%3Acolumn%29%2CorderDirection%3Adesc%2CotherBucket%3A%21t%2CparentFormat%3A%28id%3Aterms%29%2Csize%3A10%29%2Cscale%3Aordinal%2CsourceField%3Acategory.keyword%29%2Cebcb19af-0900-4439-949f-d8cd9bccde19%3A%28dataType%3Adate%2CisBucketed%3A%21t%2Clabel%3Aorder_date%2CoperationType%3Adate_histogram%2Cparams%3A%28includeEmptyRows%3A%21t%2Cinterval%3A%271d%27%29%2Cscale%3Ainterval%2CsourceField%3Aorder_date%29%29%2CincompleteColumns%3A%28%29%2CindexPatternId%3Aff959d40-b880-11e8-a6d9-e546fe2bba5f%29%29%29%29%29%2Cfilters%3A%21%28%29%2Cquery%3A%28language%3Akuery%2Cquery%3A%27%27%29%2Creferences%3A%21%28%28id%3Aff959d40-b880-11e8-a6d9-e546fe2bba5f%2Cname%3Aindexpattern-datasource-layer-97c63ea6-9305-4755-97d1-0f26817c6f9a%2Ctype%3Aindex-pattern%29%29%2CresolvedDateRange%3A%28fromDate%3A%272023-03-21T19%3A16%3A12.417Z%27%2CtoDate%3A%272023-03-28T19%3A16%3A12.417Z%27%29%2CsearchSessionId%3A%2743960930-60e1-42e3-a611-c6ae4a3af63c%27%2Cvisualization%3A%28activeId%3AlnsXY%2Cstate%3A%28axisTitlesVisibilitySettings%3A%28x%3A%21f%2CyLeft%3A%21f%2CyRight%3A%21t%29%2CfittingFunction%3ANone%2CgridlinesVisibilitySettings%3A%28x%3A%21t%2CyLeft%3A%21t%2CyRight%3A%21t%29%2Clayers%3A%21%28%28accessors%3A%21%28%275575214b-7f21-4b6c-8bc1-34433c6a0c58%27%29%2ClayerId%3A%2797c63ea6-9305-4755-97d1-0f26817c6f9a%27%2ClayerType%3Adata%2Cposition%3Atop%2CseriesType%3Abar_percentage_stacked%2CshowGridlines%3A%21f%2CsplitAccessor%3A%279f61a7df-198e-4754-b34c-81ed544136ba%27%2CxAccessor%3Aebcb19af-0900-4439-949f-d8cd9bccde19%29%29%2Clegend%3A%28isVisible%3A%21t%2ClegendSize%3Aauto%2Cposition%3Aright%29%2CpreferredSeriesType%3Abar_percentage_stacked%2CtickLabelsVisibilitySettings%3A%28x%3A%21t%2CyLeft%3A%21t%2CyRight%3A%21t%29%2CvalueLabels%3Ashow%2CyLeftExtent%3A%28mode%3Afull%29%2CyRightExtent%3A%28mode%3Afull%29%29%29%29%29%2CobjectType%3Alens%2Ctitle%3A%27Report%20Test%27%2Cversion%3A%278.8.0%27%29`

This was copied from a Lens that I opened from the Visualize Library, and made no changes. It jumps out to me that the URL is very long, because long URLs do cause problems in some environments.

After [decoding](https://meyerweb.com/eric/tools/dencoder/) the query string contents and parsing the [rison into JSON](https://observablehq.com/@nyurik/rison-online-decoder-encoder), the locator params come out to:
```
  "locatorParams": {
    "id": "LENS_APP_LOCATOR",
    "params": {
      "activeDatasourceId": "formBased",
      "dataViewSpecs": [],
      "datasourceStates": {
        "formBased": {
          "isLoading": false,
          "state": {
            "currentIndexPatternId": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
            "layers": {
              "97c63ea6-9305-4755-97d1-0f26817c6f9a": {
                "columnOrder": [
                  "9f61a7df-198e-4754-b34c-81ed544136ba",
                  "ebcb19af-0900-4439-949f-d8cd9bccde19",
                  "5575214b-7f21-4b6c-8bc1-34433c6a0c58"
                ],
                "columns": {
                  "5575214b-7f21-4b6c-8bc1-34433c6a0c58": {
                    "dataType": "number",
                    "isBucketed": false,
                    "label": "Count of records",
                    "operationType": "count",
                    "scale": "ratio",
                    "sourceField": "___records___"
                  },
                  "9f61a7df-198e-4754-b34c-81ed544136ba": {
                    "dataType": "string",
                    "isBucketed": true,
                    "label": "Top values of category.keyword",
                    "operationType": "terms",
                    "params": {
                      "missingBucket": false,
                      "orderBy": {
                        "columnId": "5575214b-7f21-4b6c-8bc1-34433c6a0c58",
                        "type": "column"
                      },
                      "orderDirection": "desc",
                      "otherBucket": true,
                      "parentFormat": {
                        "id": "terms"
                      },
                      "size": 10
                    },
                    "scale": "ordinal",
                    "sourceField": "category.keyword"
                  },
                  "ebcb19af-0900-4439-949f-d8cd9bccde19": {
                    "dataType": "date",
                    "isBucketed": true,
                    "label": "order_date",
                    "operationType": "date_histogram",
                    "params": {
                      "includeEmptyRows": true,
                      "interval": "1d"
                    },
                    "scale": "interval",
                    "sourceField": "order_date"
                  }
                },
                "incompleteColumns": {},
                "indexPatternId": "ff959d40-b880-11e8-a6d9-e546fe2bba5f"
              }
            }
          }
        }
      },
      "filters": [],
      "query": {
        "language": "kuery",
        "query": ""
      },
      "references": [
        {
          "id": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
          "name": "indexpattern-datasource-layer-97c63ea6-9305-4755-97d1-0f26817c6f9a",
          "type": "index-pattern"
        }
      ],
      "resolvedDateRange": {
        "fromDate": "2023-03-21T19:16:12.417Z",
        "toDate": "2023-03-28T19:16:12.417Z"
      },
      "searchSessionId": "43960930-60e1-42e3-a611-c6ae4a3af63c",
      "visualization": {
        "activeId": "lnsXY",
        "state": {
          "axisTitlesVisibilitySettings": {
            "x": false,
            "yLeft": false,
            "yRight": true
          },
          "fittingFunction": "None",
          "gridlinesVisibilitySettings": {
            "x": true,
            "yLeft": true,
            "yRight": true
          },
          "layers": [
            {
              "accessors": [
                "5575214b-7f21-4b6c-8bc1-34433c6a0c58"
              ],
              "layerId": "97c63ea6-9305-4755-97d1-0f26817c6f9a",
              "layerType": "data",
              "position": "top",
              "seriesType": "bar_percentage_stacked",
              "showGridlines": false,
              "splitAccessor": "9f61a7df-198e-4754-b34c-81ed544136ba",
              "xAccessor": "ebcb19af-0900-4439-949f-d8cd9bccde19"
            }
          ],
          "legend": {
            "isVisible": true,
            "legendSize": "auto",
            "position": "right"
          },
          "preferredSeriesType": "bar_percentage_stacked",
          "tickLabelsVisibilitySettings": {
            "x": true,
            "yLeft": true,
            "yRight": true
          },
          "valueLabels": "show",
          "yLeftExtent": {
            "mode": "full"
          },
          "yRightExtent": {
            "mode": "full"
          }
        }
      }
    }
```

For a saved Lens with no changes, we need a more optimized form of `locatorParams`, since they go into a URL that is often handled outside of Kibana. The data in the locatorParams should be simplified and reflect the saved object ID backing the Lens:
```
  "locatorParams": {
    "id": "LENS_APP_LOCATOR",
    "params": {
      "savedObjectId": "f9634c20-cd9c-11ed-a8c8-4bb7d8009a23",
      "resolvedDateRange": {
        "fromDate": "2023-03-21T19:16:12.417Z",
        "toDate": "2023-03-28T19:16:12.417Z"
      },
    }
```

This shortens the URL which is great for several reasons:
 * Long URLs cause problems in Watcher and with many proxies
 * In many use cases, users deploy a custom script can get a report from any saved object in a generic way, just by plugging in a different saved object ID into the POST URL. They can use other Kibana APIs to get a list of all saved object IDs they're interested in.
 * Users have told us, when they have a Watch generate reports on a saved object, when they edit the saved object they expect the new reports to reflect the changes.

For all these reasons, we should avoid having the locatorParams contain just the full internal state of what is stored in the saved object. We should make it as referential as it can be, by using the `savedObjectId` field of the `LensAppLocatorParams`

Reply:
Fixed in https://github.com/elastic/kibana/pull/153429/commits/fa61289b5b65c1b5bcfd5e3bca8ccba92671f016